File: sdk/python-sdk/verity_sdk/protocols/v1_0/IssueCredential.py
from typing import Dict

from verity_sdk.protocols.Protocol import Protocol
from verity_sdk.utils import COMMUNITY_MSG_QUALIFIER
from verity_sdk.utils.Exeptions import WrongSetupException


class IssueCredential(Protocol):
    """
    The IssueCredential protocol allows one self-sovereign party to issue a verifiable credential to another
    self-sovereign party. On the verity-application, these protocols use `anoncreds` as defined in the
    Hyperledger-indy project. In the verity-sdk, the interface presented by these objects allow key-value attributes
    to be issued to parties that have a DID-comm channel.
    """
    MSG_FAMILY = 'issue-credential'
    """the family name for the message family"""
    MSG_FAMILY_VERSION = '1.0'
    """the version for the message family"""

    PROPOSE = 'propose'
    """Name for 'propose' control message"""
    OFFER = 'offer'
    """Name for 'offer' control message"""
    REQUEST = 'request'
    """Name for 'request' control message"""
    ISSUE = 'issue'
    """Name for 'issue' control message"""
    REJECT = 'reject'
    """Name for 'reject' control message"""
    STATUS = 'status'
    """Name for 'status' control message"""

    SENT = 'sent'
    """Name for 'sent' signal message"""
    ACCEPT_REQUEST = 'accept-request'
    """Name for 'accept-request' signal message"""
    PROTOCOL_INVITATION = 'protocol-invitation'
    """Name for 'protocol-invitation' signal message"""

    #pylint: disable=too-many-arguments
    def __init__(self,
                 for_relationship: str,
                 thread_id: str = None,
                 cred_def_id: str = None,
                 values: Dict[str, str] = None,
                 comment: str = None,
                 price: str = '0',
                 auto_issue: bool = False,
                 by_invitation: bool = False):
        """
        Args:
            for_relationship (str): the relationship identifier (DID) for the pairwise relationship that will be used
            thread_id (str): The thread id of the already started protocol
            cred_def_id (str): the Credential Definition that will be used to issue the credential
            values (Dict[str, str]): a map of key-value pairs that make up the attributes in the credential
            comment (str): a human readable comment that is presented before issuing the credential
            price (str): (possible future feature) can and should be left un-set
            auto_issue (bool): flag to automatically issue the credential after receiving response from the receiver (skip getting
                signal for the credential request and waiting for the issue control message)
            by_invitation (bool): flag to create out-of-band invitation as a part of the IssueCredential protocol

        """
        super().__init__(
            self.MSG_FAMILY,
            self.MSG_FAMILY_VERSION,
            msg_qualifier=COMMUNITY_MSG_QUALIFIER,
            thread_id=thread_id
        )

        self.created = thread_id is None
        self.for_relationship = for_relationship
        self.cred_def_id = cred_def_id
        self.values = values
        self.comment = comment
        self.price = '0'  # price should be 0 regardless of what is set (until we support price)
        self.auto_issue = auto_issue
        self.by_invitation = by_invitation

    async def propose_credential(self, context):
        """
        Directs verity-application to send credential proposal.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.propose_credential_msg_packed(context))

    def propose_credential_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        if not self.created:
            raise WrongSetupException('Unable to propose credentials when NOT starting the interaction')

        msg = self._get_base_message(self.PROPOSE)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        msg['cred_def_id'] = self.cred_def_id
        msg['credential_values'] = self.values
        msg['comment'] = self.comment
        return msg

    async def propose_credential_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.propose_credential_msg())

    async def offer_credential(self, context):
        """
        Directs verity-application to send credential offer.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.offer_credential_msg_packed(context))

    def offer_credential_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        if not self.created:
            raise WrongSetupException('Unable to offer credentials when NOT starting the interaction')

        msg = self._get_base_message(self.OFFER)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        msg['cred_def_id'] = self.cred_def_id
        msg['credential_values'] = self.values
        msg['comment'] = self.comment
        msg['price'] = self.price
        msg['auto_issue'] = self.auto_issue
        msg['by_invitation'] = self.by_invitation
        return msg

    async def offer_credential_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.offer_credential_msg())

    async def request_credential(self, context):
        """
        Directs verity-application to send credential request.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.request_credential_msg_packed(context))

    def request_credential_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.REQUEST)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        msg['cred_def_id'] = self.cred_def_id
        msg['comment'] = self.comment
        return msg

    async def request_credential_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.request_credential_msg())

    async def issue_credential(self, context):
        """
        Directs verity-application to issue credential and send it

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.issue_credential_msg_packed(context))

    def issue_credential_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.ISSUE)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        msg['comment'] = self.comment
        return msg

    async def issue_credential_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.issue_credential_msg())

    async def reject(self, context):
        """
        Directs verity-application to reject the credential protocol

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.reject_msg_packed(context))

    def reject_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.REJECT)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        msg['comment'] = self.comment
        return msg

    async def reject_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.reject_msg())

    async def status(self, context):
        """
        Ask for status from the verity-application agent

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent
        """
        await self.send_message(context, await self.status_msg_packed(context))

    def status_msg(self):
        """
        Creates the control message without packaging and sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the constructed message (dict object)
        """
        msg = self._get_base_message(self.STATUS)
        self._add_thread(msg)
        self._add_relationship(msg, self.for_relationship)
        return msg

    async def status_msg_packed(self, context):
        """
        Creates and packages message without sending it.

        Args:
            context (Context): an instance of the Context object initialized to a verity-application agent

        Return:
            the bytes ready for transport
        """
        return await self.get_message_bytes(context, self.status_msg())

File: sdk/nodejs-sdk/src/protocols/v1_0/IssueCredential.js
const Protocol = require('../Protocol')
const utils = require('../../utils')
/**
 * An interface for controlling a 1.0 IssueCredential protocol.
 * @extends Protocol
 * @see <a href="https://github.com/hyperledger/aries-rfcs/tree/bb42a6c35e0d5543718fb36dd099551ab192f7b0/features/0036-issue-credential" target="_blank" rel="noopener noreferrer">Aries 0036: Issue Credential Protocol 1.0</a>
 */
class IssueCredentialV10 extends Protocol {
  /**
   * Creates an IssueCredentialV10 Protocol
   * @param forRelationship the relationship identifier (DID) for the pairwise relationship that will be used
   * @param threadId the thread id of the already started protocol
   * @param credDefId the Credential Definition that will be used to issue the credential
   * @param values a map of key-value pairs that make up the attributes in the credential
   * @param comment a human readable comment that is presented before issuing the credential
   * @param price token price (NOT CURRENTLY USED)
   * @param autoIssue flag to automatically issue the credential after receiving response from the receiver (skip getting
   *                  signal for the credential request and waiting for the issue control message)
   * @param byInvitation flag to create out-of-band invitation as a part of the IssueCredential protocol
   * @return 1.0 IssueCredential object
   *
   * @property {String} msgFamily - 'issue-credential'
   * @property {String} msgFamilyVersion - '1.0'
   * @property {String} msgQualifier - 'Community Qualifier'
   * @property {String} this.msgNames.OFFER_CREDENTIAL - 'offer'
   * @property {String} this.msgNames.PROPOSE_CREDENTIAL - 'proposal'
   * @property {String} this.msgNames.ISSUE_CREDENTIAL - 'issue'
   * @property {String} this.msgNames.REQUEST_CREDENTIAL - 'request'
   * @property {String} this.msgNames.REJECT_CREDENTIAL - 'reject'
   * @property {String} this.msgNames.CREDENTIAL_STATUS - 'status'
   * @property {String} this.msgNames.SENT - 'sent'
   * @property {String} this.msgNames.ACCEPT_REQUEST - 'accept-request'
   * @property {String} this.msgNames.PROTOCOL_INVITATION - 'protocol-invitation'
   */
  constructor (forRelationship,
    threadId = null,
    credDefId = '',
    values = {},
    comment = '',
    price = 0,
    autoIssue = false,
    byInvitation = false) {
    const msgFamily = 'issue-credential'
    const msgFamilyVersion = '1.0'
    const msgQualifier = utils.constants.COMMUNITY_MSG_QUALIFIER
    super(msgFamily, msgFamilyVersion, msgQualifier, threadId)

    this.forRelationship = forRelationship
    this.credDefId = credDefId
    this.values = values
    this.comment = comment
    this.price = price
    this.autoIssue = autoIssue
    this.byInvitation = byInvitation
    this.created = threadId === null

    this.msgNames.OFFER_CREDENTIAL = 'offer'
    this.msgNames.PROPOSE_CREDENTIAL = 'proposal'
    this.msgNames.ISSUE_CREDENTIAL = 'issue'
    this.msgNames.REQUEST_CREDENTIAL = 'request'
    this.msgNames.REJECT_CREDENTIAL = 'reject'
    this.msgNames.CREDENTIAL_STATUS = 'status'
    this.msgNames.SENT = 'sent'
    this.msgNames.ACCEPT_REQUEST = 'accept-request'
    this.msgNames.PROTOCOL_INVITATION = 'protocol-invitation'
  }

  /**
     * Directs verity-application to send credential offer.
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async offerCredential (context) {
    await this.sendMessage(context, await this.offerCredentialMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #offerCredential
     */
  async offerCredentialMsgPacked (context) {
    return this.getMessageBytes(context, this.offerCredentialMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #offerCredential
     */
  offerCredentialMsg () {
    if (!this.created) {
      throw new utils.WrongSetupError('Unable to offer credentials when NOT starting the interaction')
    }

    var msg = this._getBaseMessage(this.msgNames.OFFER_CREDENTIAL)
    msg['~for_relationship'] = this.forRelationship
    msg.cred_def_id = this.credDefId
    msg.comment = this.comment
    msg.price = this.price
    msg.credential_values = this.values
    msg.auto_issue = this.autoIssue
    msg.by_invitation = this.byInvitation
    msg = this._addThread(msg)

    return msg
  }

  /**
     * Directs verity-application to send credential proposal.
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async proposeCredential (context) {
    await this.sendMessage(context, await this.proposeCredentialMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #proposeCredential
     */
  async proposeCredentialMsgPacked (context) {
    return this.getMessageBytes(context, this.proposeCredentialMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #proposeCredential
     */
  proposeCredentialMsg () {
    if (!this.created) {
      throw new utils.WrongSetupError('Unable to propose credentials when NOT starting the interaction')
    }

    var msg = this._getBaseMessage(this.msgNames.PROPOSE_CREDENTIAL)
    msg['~for_relationship'] = this.forRelationship
    msg.cred_def_id = this.credDefId
    msg.comment = this.comment
    msg.credential_values = this.values
    msg = this._addThread(msg)

    return msg
  }

  /**
     * Directs verity-application to send credential request.
     *
     * @param context
     */
  async requestCredential (context) {
    await this.sendMessage(context, await this.requestCredentialMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #requestCredential
     */
  async requestCredentialMsgPacked (context) {
    return this.getMessageBytes(context, this.requestCredentialMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #requestCredential
     */
  requestCredentialMsg () {
    var msg = this._getBaseMessage(this.msgNames.REQUEST_CREDENTIAL)
    msg['~for_relationship'] = this.forRelationship
    msg.cred_def_id = this.credDefId
    msg.comment = this.comment
    msg = this._addThread(msg)

    return msg
  }

  /**
     * Directs verity-application to issue credential and send it
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async issueCredential (context) {
    await this.sendMessage(context, await this.issueCredentialMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #issueCredential
     */
  async issueCredentialMsgPacked (context) {
    return this.getMessageBytes(context, this.issueCredentialMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #issueCredential
     */
  issueCredentialMsg () {
    var msg = this._getBaseMessage(this.msgNames.ISSUE_CREDENTIAL)
    msg['~for_relationship'] = this.forRelationship
    msg.comment = this.comment
    msg = this._addThread(msg)
    return msg
  }

  /**
     * Directs verity-application to reject the credential protocol
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async reject (context) {
    await this.sendMessage(context, await this.rejectMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #reject
     */
  async rejectMsgPacked (context) {
    return this.getMessageBytes(context, this.rejectMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #reject
     */
  rejectMsg () {
    var msg = this._getBaseMessage(this.msgNames.REJECT_CREDENTIAL)
    msg['~for_relationship'] = this.forRelationship
    msg.comment = this.comment
    msg = this._addThread(msg)

    return msg
  }

  /**
     * Ask for status from the verity-application agent
     *
     * @param context an instance of the Context object initialized to a verity-application agent
     */
  async status (context) {
    await this.sendMessage(context, await this.statusMsgPacked(context))
  }

  /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     *
     * @see #status
     */
  async statusMsgPacked (context) {
    return this.getMessageBytes(context, this.statusMsg())
  }

  /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     *
     * @see #status
     */
  statusMsg () {
    var msg = this._getBaseMessage(this.msgNames.CREDENTIAL_STATUS)
    msg['~for_relationship'] = this.forRelationship
    msg = this._addThread(msg)
    return msg
  }
}
module.exports = IssueCredentialV10
